cou/steps/analyze.py

```diff
@@ -0,0 +1,205 @@
+        except KeyError:
+            logging.warning("package not found for application: %s", self.name)
+            pkg_name = ""
+        return pkg_name
+
+    # NOTE (gabrielcocenza) Ideally, the application should provide the openstack version
+    # and packages versions by a charm action. This might be possible with Sunbeam.
+    async def _get_current_os_versions(self, unit: str) -> str:
+        """Get the openstack version of a unit."""
+        version = ""
+        pkg_version = self._get_pkg_version(unit)
+        self.units[unit]["pkg_version"] = pkg_version
+        self.pkg_version_units[pkg_version].add(unit)
+
+        # for openstack releases >= wallaby
+        codename = await self._get_openstack_release(unit, model_name=self.model_name)
+        if codename:
+            version = codename
+        # for openstack releases < wallaby
+        elif self.pkg_name and pkg_version:
+            version = get_os_code_info(self.pkg_name, pkg_version)
+        return version
+
+    def _get_os_origin(self) -> str:
+        """Get application configuration for openstack-origin or source."""
+        for origin in ("openstack-origin", "source"):
+            if self.config.get(origin):
+                return self.config[origin].get("value", "")
+
+        logging.warning("Failed to get origin for %s, no origin config found", self.name)
+        return ""
+
+    async def _get_openstack_release(
+        self, unit: str, model_name: Union[str, None] = None
+    ) -> Union[str, None]:
+        """Return the openstack release codename based on /etc/openstack-release."""
+        cmd = "grep -Po '(?<=OPENSTACK_CODENAME=).*' /etc/openstack-release"
+        try:
+            out = await async_run_on_unit(unit, cmd, model_name=model_name, timeout=20)
+        except CommandRunFailed:
+            logging.warning("Fall back to version check for OpenStack codename")
+            return None
+        return out["Stdout"].strip()
+
+    def _get_pkg_version(self, unit: str) -> str:
+        """Get the openstack package version in a unit."""
+        try:
+            return self.status.units[unit].workload_version
+        except AttributeError:
+            logging.warning("Failed to get pkg version for '%s'", self.name)
+            return ""
+
+
+async def generate_model() -> set[Application]:
+    """Generate the applications model."""
+    juju_status = await async_get_full_juju_status()
+    model_name = juju_status.model.name
+    apps = {
+        await Application(
+            name=app,
+            status=app_status,
+            config=await async_get_application_config(app),
+            model_name=model_name,
+        ).fill()
+        for app, app_status in juju_status.applications.items()
+    }
+    # NOTE(gabrielcocenza) Not all openstack-charms are mapped in the zaza lookup.
+    openstack_apps = {app for app in apps if app.charm in CHARM_TYPES}
+    not_supported_apps = apps - openstack_apps
```

Comment:
I think we should not filter on the Analyze part at all to say the truth. This might be a task for the planning part because we might want interact with non-openstack apps (hacluster, rabbitmq, mysql and etc) in a close future.